src/pydicom/filebase.py

```diff
@@ -1,211 +1,332 @@
+            bytes_read += self.read(length - num_bytes)
+            attempts += 1
+
+        if num_bytes == length:
+            return bytes(bytes_read)
+
+        # Failed to read the required length
+        raise EOFError(
+            f"Unexpected end of file. Read {num_bytes} bytes of {length} "
+            f"expected starting at position 0x{self.tell() - num_bytes:x}"
+        )
+
+    def read_tag(self) -> tuple[int, int]:
+        """Return a DICOM tag value read from the buffer."""
+        return cast(
+            tuple[int, int],
+            self._tag_unpacker(self.read_exact(4)),
+        )
+
+    def read_UL(self) -> int:
+        """Return a UL value read from the buffer."""
+        return cast(int, self._ul_unpacker(self.read(4))[0])
+
+    def read_US(self) -> int:
+        """Return a US value read from the buffer."""
+        return cast(int, self._us_unpacker(self.read(2))[0])
+
+    def seek(self, offset: int, whence: int = os.SEEK_SET, /) -> int:
+        """Change the buffer position to the given byte `offset`, relative to
+        the position indicated by `whence` and return the new absolute position.
+        """
+        name = type(self).__name__
+        buffer = type(self.parent).__name__
+        raise OSError(
+            f"'{name}' cannot be used with '{buffer}': object has no seek() method"
+        )
+
+    @property
+    def supports_read(self) -> bool:
+        """Return ``True`` if :attr:`~pydicom.filebase.DicomIO.parent` has
+        'read', 'tell' and 'seek' methods.
+        """
+        methods = {
+            x for x in dir(self.parent) if callable(getattr(self.parent, x, None))
+        }
+
+        return {"read", "tell", "seek"} <= methods
+
+    @property
+    def supports_write(self) -> bool:
+        """Return ``True`` if :attr:`~pydicom.filebase.DicomIO.parent` has
+        'write', 'tell' and 'seek' methods.
+        """
+        methods = {
+            x for x in dir(self.parent) if callable(getattr(self.parent, x, None))
+        }
+
+        return {"write", "tell", "seek"} <= methods
+
+    def tell(self) -> int:
+        """Return the current stream position of the buffer"""
+        name = type(self).__name__
+        buffer = type(self.parent).__name__
+        raise OSError(
+            f"'{name}' cannot be used with '{buffer}': object has no tell() method"
+        )
+
+    def write(self, b: bytes | bytearray | memoryview) -> int:
+        """Write the bytes-like object `b` to the buffer and return the number
+        of bytes written.
+        """
+        name = type(self).__name__
+        buffer = type(self.parent).__name__
+        raise OSError(
+            f"'{name}' cannot be used with '{buffer}': object has no write() method"
+        )
+
+    def write_tag(self, tag: int) -> None:
+        """Write a DICOM tag to the buffer."""
+        self.write(self._tag_packer(tag >> 16, tag & 0xFFFF))
```

Comment:
I think this is about twice as fast as before, the type check and `Tag()` call is relatively expensive